tests/test_re.py

```diff
@@ -0,0 +1,520 @@
+    def __init__(self, char, unicat, polarity_cat):
+        self.char = char
+        self.unicat = unicat
+        self.polarity_cat = polarity_cat
+
+    def matching_string(self, draw, state):
+        if self.polarity_cat:
+            return draw(st.characters(whitelist_categories=self.unicat))
+        return draw(st.characters(blacklist_categories=self.unicat))
+
+    def non_matching_string(self, draw, state):
+        if self.polarity_cat:
+            return draw(st.characters(blacklist_categories=self.unicat))
+        return draw(st.characters(whitelist_categories=self.unicat))
+
+    def build_re(self):
+        return "\\" + self.char
+
+    @staticmethod
+    def make(draw):
+        # XXX can add more
+        return CharClass(
+            *draw(st.sampled_from([("d", ["Nd"], True), ("D", ["Nd"], False)]))
+        )
+
+
+class Dot(Re):
+    def matching_string(self, draw, state):
+        return draw(st.characters(blacklist_characters="\n"))
+
+    def non_matching_string(self, draw, state):
+        return "\n"
+
+    def build_re(self):
+        return "."
+
+    @staticmethod
+    def make(draw):
+        return Dot()
+
+
+class Escape(Re):
+    def __init__(self, c):
+        self.c = c
+
+    def matching_string(self, draw, state):
+        return self.c
+
+    def non_matching_string(self, draw, state):
+        return draw(st.characters(blacklist_characters=self.c))
+
+    def build_re(self):
+        return "\\" + self.c
+
+    @staticmethod
+    def make(draw):
+        # XXX many more escapes
+        c = draw(st.sampled_from(special_characters))
+        return Escape(c)
+
+
+class Charset(Re):
+    def __init__(self, elements):
+        # either characters, or (start, stop) tuples
+        self.elements = elements
+
+    def matching_string(self, draw, state):
+        x = draw(st.sampled_from(self.elements))
+        if isinstance(x, tuple):
+            return draw(st.characters(min_codepoint=ord(x[0]), max_codepoint=ord(x[1])))
+        return x
+
+    def non_matching_string(self, draw, state):
+        if not any(isinstance(x, tuple) for x in self.elements):
+            # easy case, only chars
+            return draw(st.characters(blacklist_characters=self.elements))
+        chars = "".join(x for x in self.elements if not isinstance(x, tuple))
+        range_stops = [ord(x[1]) for x in self.elements if isinstance(x, tuple)]
+        max_stop = max(range_stops)
+        res = draw(
+            st.characters(min_codepoint=max_stop + 1, blacklist_characters=chars)
+        )
+        return res
+
+    def build_re(self):
+        res = []
+        for x in self.elements:
+            if isinstance(x, tuple):
+                res.append(f"{x[0]}-{x[1]}")
+            else:
+                res.append(x)
+        return "[" + "".join(res) + "]"
+
+    @staticmethod
+    def make(draw):
+        # XXX character classes
+        elements = []
+        for _ in range(draw(st.integers(min_value=2, max_value=20))):
```

Comment:
This is a really tempting pattern, but for internal performance (both speed and quality) reasons it's best to use the `lists()` strategy.  For example:

```python
char = st.characters(blacklist_characters="-^]\\")
elem = char | st.tuples(char, char).filter(lambda t: t[0] != t[1]).map(sorted).map(tuple)
return Charset(draw(st.lists(elem, min_size=2, max_size=20))
```

Admittedly the `elem` definition is a little clumsy; you might perfer to use a `@st.composite` strategy for that bit.

Reply:
makes sense! I used a composite indeed